File: suites/15_special_functions/comparisons/special_functions.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>

static const double PI = 3.14159265358979323846;

struct FourierResult {
    double energy;
    double total_re;
    double total_im;
};

FourierResult fourier_series_benchmark(int num_harmonics = 16, int num_samples = 1024, int num_evals = 256) {
    double dx = 2.0 * PI / (double)num_samples;
    int num_coeffs = 2 * num_harmonics + 1;
    std::vector<double> cn_re(num_coeffs, 0.0);
    std::vector<double> cn_im(num_coeffs, 0.0);

    for (int n = -num_harmonics; n <= num_harmonics; ++n) {
        int idx = n + num_harmonics;
        double sum_re = 0.0;
        double sum_im = 0.0;
        for (int k = 0; k < num_samples; ++k) {
            double x = (double)k * dx;
            double fx = x * (2.0 * PI - x) * (x - PI);
            double angle = (double)n * x;
            double cos_a = std::cos(angle);
            double sin_a = std::sin(angle);
            sum_re += fx * cos_a;
            sum_im += -fx * sin_a;
        }
        cn_re[idx] = sum_re / (double)num_samples;
        cn_im[idx] = sum_im / (double)num_samples;
    }

    double d_eval = 2.0 * PI / (double)num_evals;
    double total_re = 0.0;
    double total_im = 0.0;
    for (int e = 0; e < num_evals; ++e) {
        double x_ev = (double)e * d_eval;
        double re_recon = 0.0;
        double im_recon = 0.0;
        for (int n = -num_harmonics; n <= num_harmonics; ++n) {
            int idx = n + num_harmonics;
            double cre = cn_re[idx];
            double cim = cn_im[idx];
            double angle = (double)n * x_ev;
            double e_re = std::cos(angle);
            double e_im = std::sin(angle);
            re_recon += cre * e_re - cim * e_im;
            im_recon += cre * e_im + cim * e_re;
        }
        total_re += re_recon;
        total_im += im_recon;
    }

    double energy = 0.0;
    for (int i = 0; i < num_coeffs; ++i) {
        energy += cn_re[i] * cn_re[i] + cn_im[i] * cn_im[i];
    }

    return {energy, total_re, total_im};
}

struct HyperResult {
    double sum_re;
    double sum_im;
    int total_terms;
};

HyperResult hypergeometric_2f1_benchmark(int grid_size = 50, int max_terms = 200, double tol = 1e-15) {
    double a = 0.5;
    double b = 1.0;
    double c = 2.0;
    double total_re = 0.0;
    double total_im = 0.0;
    int total_terms = 0;
    double tol2 = tol * tol;

    for (int ix = 0; ix < grid_size; ++ix) {
        double zr = ((double)ix - 25.0) * 0.015;
        for (int iy = 0; iy < grid_size; ++iy) {
            double zi = ((double)iy - 25.0) * 0.015;

            double term_re = 1.0;
            double term_im = 0.0;
            double sum_re = 1.0;
            double sum_im = 0.0;

            int n_used = max_terms;
            for (int n = 1; n < max_terms; ++n) {
                double dn = (double)n;
                double factor_scalar = ((a + dn - 1.0) * (b + dn - 1.0)) / ((c + dn - 1.0) * dn);
                double fz_re = factor_scalar * zr;
                double fz_im = factor_scalar * zi;

                double new_t_re = term_re * fz_re - term_im * fz_im;
                double new_t_im = term_re * fz_im + term_im * fz_re;
                term_re = new_t_re;
                term_im = new_t_im;

                sum_re += term_re;
                sum_im += term_im;

                double mag2 = term_re * term_re + term_im * term_im;
                if (mag2 < tol2) {
                    n_used = n;
                    break;
                }
            }
            total_terms += n_used;
            total_re += sum_re;
            total_im += sum_im;
        }
    }

    return {total_re, total_im, total_terms};
}

int main() {
    // 1. Fourier Series
    auto t0 = std::chrono::high_resolution_clock::now();
    FourierResult f_res = fourier_series_benchmark();
    auto t1 = std::chrono::high_resolution_clock::now();
    double fourier_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    // 2. Hypergeometric Series
    auto t2 = std::chrono::high_resolution_clock::now();
    HyperResult h_res = hypergeometric_2f1_benchmark();
    auto t3 = std::chrono::high_resolution_clock::now();
    double hyper_ms = std::chrono::duration<double, std::milli>(t3 - t2).count();

    std::cout << "Fourier Series: energy=" << f_res.energy << ", total_re=" << f_res.total_re << " in " << fourier_ms << " ms" << std::endl;
    std::cout << "Hypergeometric 2F1: sum_re=" << h_res.sum_re << ", sum_im=" << h_res.sum_im << ", terms=" << h_res.total_terms << " in " << hyper_ms << " ms" << std::endl;

    long long chk_fourier = (long long)(f_res.energy * 100000.0) + (long long)(std::abs(f_res.total_re) * 1000.0);
    long long chk_hyper = (long long)(h_res.sum_re * 1000.0) + (long long)(h_res.sum_im * 1000.0) + (long long)h_res.total_terms;
    long long combined_checksum = (chk_fourier * 1000003LL) ^ chk_hyper;
    std::cout << "COMBINED_CHECKSUM: " << combined_checksum << std::endl;

    return 0;
}

File: suites/15_special_functions/comparisons/special_functions.rs
use std::time::Instant;

const PI: f64 = std::f64::consts::PI;

struct FourierResult {
    energy: f64,
    total_re: f64,
    total_im: f64,
}

fn fourier_series_benchmark(num_harmonics: i32, num_samples: usize, num_evals: usize) -> FourierResult {
    let dx = 2.0 * PI / (num_samples as f64);
    let num_coeffs = (2 * num_harmonics + 1) as usize;
    let mut cn_re = vec![0.0; num_coeffs];
    let mut cn_im = vec![0.0; num_coeffs];

    for n in -num_harmonics..=num_harmonics {
        let idx = (n + num_harmonics) as usize;
        let mut sum_re = 0.0;
        let mut sum_im = 0.0;
        for k in 0..num_samples {
            let x = (k as f64) * dx;
            let fx = x * (2.0 * PI - x) * (x - PI);
            let angle = (n as f64) * x;
            let cos_a = angle.cos();
            let sin_a = angle.sin();
            sum_re += fx * cos_a;
            sum_im += -fx * sin_a;
        }
        cn_re[idx] = sum_re / (num_samples as f64);
        cn_im[idx] = sum_im / (num_samples as f64);
    }

    let d_eval = 2.0 * PI / (num_evals as f64);
    let mut total_re = 0.0;
    let mut total_im = 0.0;
    for e in 0..num_evals {
        let x_ev = (e as f64) * d_eval;
        let mut re_recon = 0.0;
        let mut im_recon = 0.0;
        for n in -num_harmonics..=num_harmonics {
            let idx = (n + num_harmonics) as usize;
            let cre = cn_re[idx];
            let cim = cn_im[idx];
            let angle = (n as f64) * x_ev;
            let e_re = angle.cos();
            let e_im = angle.sin();
            re_recon += cre * e_re - cim * e_im;
            im_recon += cre * e_im + cim * e_re;
        }
        total_re += re_recon;
        total_im += im_recon;
    }

    let mut energy = 0.0;
    for i in 0..num_coeffs {
        energy += cn_re[i] * cn_re[i] + cn_im[i] * cn_im[i];
    }

    FourierResult { energy, total_re, total_im }
}

struct HyperResult {
    sum_re: f64,
    sum_im: f64,
    total_terms: i64,
}

fn hypergeometric_2f1_benchmark(grid_size: i32, max_terms: i32, tol: f64) -> HyperResult {
    let a = 0.5;
    let b = 1.0;
    let c = 2.0;
    let mut total_re = 0.0;
    let mut total_im = 0.0;
    let mut total_terms = 0;
    let tol2 = tol * tol;

    for ix in 0..grid_size {
        let zr = ((ix as f64) - 25.0) * 0.015;
        for iy in 0..grid_size {
            let zi = ((iy as f64) - 25.0) * 0.015;

            let mut term_re = 1.0;
            let mut term_im = 0.0;
            let mut sum_re = 1.0;
            let mut sum_im = 0.0;

            let mut n_used = max_terms;
            for n in 1..max_terms {
                let dn = n as f64;
                let factor_scalar = ((a + dn - 1.0) * (b + dn - 1.0)) / ((c + dn - 1.0) * dn);
                let fz_re = factor_scalar * zr;
                let fz_im = factor_scalar * zi;

                let new_t_re = term_re * fz_re - term_im * fz_im;
                let new_t_im = term_re * fz_im + term_im * fz_re;
                term_re = new_t_re;
                term_im = new_t_im;

                sum_re += term_re;
                sum_im += term_im;

                let mag2 = term_re * term_re + term_im * term_im;
                if mag2 < tol2 {
                    n_used = n;
                    break;
                }
            }
            total_terms += n_used as i64;
            total_re += sum_re;
            total_im += sum_im;
        }
    }

    HyperResult { sum_re: total_re, sum_im: total_im, total_terms }
}

fn main() {
    let t0 = Instant::now();
    let f_res = fourier_series_benchmark(16, 1024, 256);
    let fourier_ms = t0.elapsed().as_secs_f64() * 1000.0;

    let t1 = Instant::now();
    let h_res = hypergeometric_2f1_benchmark(50, 200, 1e-15);
    let hyper_ms = t1.elapsed().as_secs_f64() * 1000.0;

    println!("Fourier Series: energy={:.6}, total_re={:.6} in {:.4} ms", f_res.energy, f_res.total_re, fourier_ms);
    println!("Hypergeometric 2F1: sum_re={:.6}, sum_im={:.6}, terms={} in {:.4} ms", h_res.sum_re, h_res.sum_im, h_res.total_terms, hyper_ms);

    let chk_fourier = (f_res.energy * 100000.0) as i64 + (f_res.total_re.abs() * 1000.0) as i64;
    let chk_hyper = (h_res.sum_re * 1000.0) as i64 + (h_res.sum_im * 1000.0) as i64 + h_res.total_terms;
    let combined_checksum = (chk_fourier * 1000003) ^ chk_hyper;
    println!("COMBINED_CHECKSUM: {}", combined_checksum);
}
